lib/partition_healing/healer.js

```diff
@@ -0,0 +1,208 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var _ = require('underscore');
+var async = require('async');
+
+var Member = require('../membership/member');
+var PingSender = require('../gossip/ping-sender');
+
+/**
+ *
+ * Healer implements an algorithm to heal a partitioned ringpop cluster.
+ *
+ * @param ringpop
+ *
+ * @constructor
+ */
+function Healer(ringpop) {
+    this.ringpop = ringpop;
+    this.logger = this.ringpop.loggerFactory.getLogger('healer');
+}
+
+
+/**
+ * Attempt a heal between the current and target node. Calling this function does not
+ * result in a heal ff there are nodes that need to be reincarnated to take precedence
+ * over the faulty declarations. A cluster may need some time and multiple heal attempts
+ * before it is successfully healed.
+ *
+ * @param {string} target The address of the target node of the heal-attempt.
+ * @param {Healer~attemptHealCallback} callback called when the heal attempt is done.
+ * @protected
+ */
+Healer.prototype.attemptHeal = function attemptHeal(target, callback) {
+    var self = this;
+
+    self.ringpop.stat('increment', 'heal.attempt');
+    this.logger.info('ringpop attempt heal', {
+        local: self.ringpop.whoami(),
+        target: target
+    });
+
+    var remoteMembership = null;
+    
+    async.waterfall([
+        sendJoinRequest,
+        generateChanges,
+        processChanges
+    ], function done(err) {
+        if (err) {
+            callback(err);
+            return;
+        }
+
+        var pingableHosts = _.chain(remoteMembership).filter(function isPingable(change) {
+            return Member.isStatusPingable(change.status);
+        }).pluck('address').value();
+
+        callback(null, pingableHosts);
+    });
+
+    function sendJoinRequest(next) {
+        self.ringpop.client.protocolJoin({
+            host: target,
+            retryLimit: self.ringpop.config.get('tchannelRetryLimit'),
+            timeout: 1000
+        }, {
+            app: self.ringpop.app,
+            source: self.ringpop.whoami(),
+            incarnationNumber: self.ringpop.membership.localMember.incarnationNumber
+        }, next);
+    }
+
+    function generateChanges(joinResponse, next) {
+        remoteMembership = joinResponse.membership;
+
+        // Index local membership by address for faster lookups.
+        var localMembershipMap = _.indexBy(self.ringpop.dissemination.fullSync(), 'address');
+
+        var localChanges = [];
```

Comment:
I though this was actually easier to read because localChanges are applied to the local change and remoteChanges to the target (so the remote node) but it might make sense to keep them in sync.